subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/DefaultTaskClassInfoStore.java

```diff
@@ -97,12 +98,16 @@ private static TaskActionFactory createTaskAction(Class<? extends Task> taskType
 
         TaskActionFactory taskActionFactory;
         if (parameterTypes.length == 1) {
-            if (!parameterTypes[0].equals(IncrementalTaskInputs.class)) {
+            Class<?> parameterType = parameterTypes[0];
+            if (parameterType.equals(IncrementalTaskInputs.class)) {
+                taskActionFactory = new IncrementalTaskActionFactory(taskType, method);
+            } else if (parameterType.equals(IncrementalInputs.class)) {
+                taskActionFactory = new IncrementalActionFactory(taskType, method);
```

Comment:
I don't think this is a good name to distinguish between the two kinds of incremental inputs. We could instead use `IncrementalInputsTaskActionFactory` vs `IncrementalTaskInputsTaskActionFactory` or something.